File: backend/app/services/flood_depth.py
"""Canonical MMDA-style flood depth gauge values shared by server workflows."""

from dataclasses import dataclass
from typing import Optional

from app.models.report import ReportSeverity


@dataclass(frozen=True)
class FloodDepthMeasurement:
    """Official MMDA-correlated measurement specifications for a flood depth level."""
    key: str
    display_label: str
    inches: float
    meters: float
    centimeters: float
    formatted: str
    accessibility_class: str  # "PATV", "NPLV", "NPATV"
    severity: ReportSeverity


FLOOD_DEPTH_MEASUREMENTS: dict[str, FloodDepthMeasurement] = {
    "gutter": FloodDepthMeasurement(
        key="gutter",
        display_label="Gutter",
        inches=8.0,
        meters=0.20,
        centimeters=20.32,
        formatted='8" (0.20m)',
        accessibility_class="PATV",
        severity=ReportSeverity.LOW,
    ),
    "half-knee": FloodDepthMeasurement(
        key="half-knee",
        display_label="Half-Knee",
        inches=10.0,
        meters=0.25,
        centimeters=25.40,
        formatted='10" (0.25m)',
        accessibility_class="PATV",
        severity=ReportSeverity.LOW,
    ),
    "half-tire": FloodDepthMeasurement(
        key="half-tire",
        display_label="Half-Tire",
        inches=13.0,
        meters=0.33,
        centimeters=33.02,
        formatted='13" (0.33m)',
        accessibility_class="NPLV",
        severity=ReportSeverity.MEDIUM,
    ),
    "knee": FloodDepthMeasurement(
        key="knee",
        display_label="Knee",
        inches=19.0,
        meters=0.48,
        centimeters=48.26,
        formatted='19" (0.48m)',
        accessibility_class="NPLV",
        severity=ReportSeverity.MEDIUM,
    ),
    "tires": FloodDepthMeasurement(
        key="tires",
        display_label="Tires",
        inches=26.0,
        meters=0.66,
        centimeters=66.04,
        formatted='26" (0.66m)',
        accessibility_class="NPATV",
        severity=ReportSeverity.HIGH,
    ),
    "waist": FloodDepthMeasurement(
        key="waist",
        display_label="Waist",
        inches=37.0,
        meters=0.94,
        centimeters=93.98,
        formatted='37" (0.94m)',
        accessibility_class="NPATV",
        severity=ReportSeverity.HIGH,
    ),
    "chest": FloodDepthMeasurement(
        key="chest",
        display_label="Chest",
        inches=45.0,
        meters=1.14,
        centimeters=114.30,
        formatted='45" (1.14m)',
        accessibility_class="NPATV",
        severity=ReportSeverity.HIGH,
    ),
    "neck": FloodDepthMeasurement(
        key="neck",
        display_label="Neck & Above",
        inches=55.0,
        meters=1.40,
        centimeters=140.0,
        formatted='55"+ (1.40m+)',
        accessibility_class="NPATV",
        severity=ReportSeverity.EXTREME,
    ),
}

FLOOD_DEPTH_SEVERITIES: dict[str, ReportSeverity] = {
    k: v.severity for k, v in FLOOD_DEPTH_MEASUREMENTS.items()
}

_DEPTH_ALIASES = {
    "neck & above": "neck",
    "neck and above": "neck",
}


def normalize_flood_depth(value: Optional[str]) -> Optional[str]:
    """Return a canonical gauge ID, accepting legacy display labels at the API edge."""
    if value is None:
        return None
    normalized = " ".join(value.strip().lower().split())
    normalized = _DEPTH_ALIASES.get(normalized, normalized)
    if normalized not in FLOOD_DEPTH_MEASUREMENTS:
        raise ValueError("Depth must be a supported flood gauge level.")
    return normalized


def severity_for_flood_depth(depth: str) -> ReportSeverity:
    canonical = normalize_flood_depth(depth)
    if not canonical:
        raise ValueError("Depth must be a supported flood gauge level.")
    return FLOOD_DEPTH_MEASUREMENTS[canonical].severity


def get_flood_depth_measurement(depth: Optional[str]) -> Optional[FloodDepthMeasurement]:
    """Return the official MMDA-correlated measurement details for a depth value."""
    if not depth:
        return None
    try:
        canonical = normalize_flood_depth(depth)
        return FLOOD_DEPTH_MEASUREMENTS.get(canonical) if canonical else None
    except ValueError:
        return None


def format_flood_depth(depth: Optional[str]) -> Optional[str]:
    """Format a depth string into readable MMDA measurements, e.g. 'Knee • 19\" (0.48m)'."""
    measurement = get_flood_depth_measurement(depth)
    if not measurement:
        return depth
    return f"{measurement.display_label} • {measurement.formatted}"

File: frontend/src/lib/floodDepth.ts
export type FloodSeverity = "low" | "medium" | "high" | "extreme";

export interface FloodDepthSpec {
  id: string;
  label: string;
  severity: FloodSeverity;
  inches: number;
  meters: number;
  centimeters: number;
  formatted: string;
  description: string;
  accessibilityClass: "PATV" | "NPLV" | "NPATV";
}

export const FLOOD_DEPTH_SPECS: Record<string, FloodDepthSpec> = {
  gutter: {
    id: "gutter",
    label: "Gutter",
    severity: "low",
    inches: 8.0,
    meters: 0.20,
    centimeters: 20.32,
    formatted: '8" (0.20m)',
    description: '8" (0.20m)',
    accessibilityClass: "PATV",
  },
  "half-knee": {
    id: "half-knee",
    label: "Half-Knee",
    severity: "low",
    inches: 10.0,
    meters: 0.25,
    centimeters: 25.40,
    formatted: '10" (0.25m)',
    description: '10" (0.25m)',
    accessibilityClass: "PATV",
  },
  "half-tire": {
    id: "half-tire",
    label: "Half-Tire",
    severity: "medium",
    inches: 13.0,
    meters: 0.33,
    centimeters: 33.02,
    formatted: '13" (0.33m)',
    description: '13" (0.33m)',
    accessibilityClass: "NPLV",
  },
  knee: {
    id: "knee",
    label: "Knee",
    severity: "medium",
    inches: 19.0,
    meters: 0.48,
    centimeters: 48.26,
    formatted: '19" (0.48m)',
    description: '19" (0.48m)',
    accessibilityClass: "NPLV",
  },
  tires: {
    id: "tires",
    label: "Tires",
    severity: "high",
    inches: 26.0,
    meters: 0.66,
    centimeters: 66.04,
    formatted: '26" (0.66m)',
    description: '26" (0.66m)',
    accessibilityClass: "NPATV",
  },
  waist: {
    id: "waist",
    label: "Waist",
    severity: "high",
    inches: 37.0,
    meters: 0.94,
    centimeters: 93.98,
    formatted: '37" (0.94m)',
    description: '37" (0.94m)',
    accessibilityClass: "NPATV",
  },
  chest: {
    id: "chest",
    label: "Chest",
    severity: "high",
    inches: 45.0,
    meters: 1.14,
    centimeters: 114.30,
    formatted: '45" (1.14m)',
    description: '45" (1.14m)',
    accessibilityClass: "NPATV",
  },
  neck: {
    id: "neck",
    label: "Neck & Above",
    severity: "extreme",
    inches: 55.0,
    meters: 1.40,
    centimeters: 140.0,
    formatted: '55"+ (1.40m+)',
    description: '55"+ (1.40m+)',
    accessibilityClass: "NPATV",
  },
};

export const FLOOD_DEPTH_OPTIONS: FloodDepthSpec[] = [
  FLOOD_DEPTH_SPECS["gutter"],
  FLOOD_DEPTH_SPECS["half-knee"],
  FLOOD_DEPTH_SPECS["half-tire"],
  FLOOD_DEPTH_SPECS["knee"],
  FLOOD_DEPTH_SPECS["tires"],
  FLOOD_DEPTH_SPECS["waist"],
  FLOOD_DEPTH_SPECS["chest"],
  FLOOD_DEPTH_SPECS["neck"],
];

const DEPTH_ALIASES: Record<string, string> = {
  "neck-and-above": "neck",
  "neck-&-above": "neck",
  "neck and above": "neck",
  "neck & above": "neck",
};

export function getFloodDepthSpec(depth?: string | null): FloodDepthSpec | null {
  if (!depth || depth === "null") return null;
  const normalized = depth.toLowerCase().trim();
  const aliasMatch = DEPTH_ALIASES[normalized] || DEPTH_ALIASES[normalized.replace(/_/g, "-")];
  const key = aliasMatch || normalized.replace(/_/g, "-");
  return FLOOD_DEPTH_SPECS[key] ?? null;
}

export function formatFloodDepth(
  depth?: string | null,
  options?: { fallback?: string; includeMeasurement?: boolean; compact?: boolean }
): string {
  const fallback = options?.fallback ?? "Not specified";
  const includeMeasurement = options?.includeMeasurement ?? true;
  const compact = options?.compact ?? false;

  const spec = getFloodDepthSpec(depth);
  if (!spec) {
    if (!depth || depth === "null") return fallback;
    return depth.replace(/_/g, " ");
  }

  if (!includeMeasurement) return spec.label;
  if (compact) return `${spec.label} (${spec.formatted})`;
  return `${spec.label} • ${spec.formatted}`;
}
